### PR TITLE
Feature keyvault reference

### PR DESCRIPTION
**Pull request by RossMurr4y**
Implementing a number of functions to allow the creation of ARM template parameters, to address the requirement for them in codeontap/gen3-azure#72

Also includes the functions to create Deployment resources, which allow deployment-scoped resources which will be important for User accounts in the future.

### CODE REVIEW

File: components/baseline/setup.ftl [hunk 1]
```diff
@@ -55,7 +55,7 @@
     [#local baselineComponentIds = getBaselineComponentIds(
       baselineLinks,
       AZURE_CMK_RESOURCE_TYPE,
-      AZURE_SSH_PRIVATE_KEY_RESOURCE_TYPE,
+      AZURE_KEYVAULT_SECRET_RESOURCE_TYPE,
```
Comment (roleyfoley) [inline components/baseline/setup.ftl]: This one shouldn't be a resource type it should be the id of the ssh key pair resource `vmKeyPair`

Comment (roleyfoley) [reply components/baseline/setup.ftl]: To make the distinction between the id vs type of cmk I would make it "cmk" as well

Comment (RossMurr4y) [reply components/baseline/setup.ftl]: This has been addressed in 0be0928

File: components/bastion/setup.ftl [hunk 2]
```diff
@@ -0,0 +1,173 @@
...
+
+  [#if ! networkLinkTarget?has_content ]
+      [@fatal message="Network could not be found" context=networkLink /]
+      [#return]
+  [/#if]
+
+  [#local networkConfiguration = networkLinkTarget.Configuration.Solution]
+  [#local networkResources = networkLinkTarget.State.Resources ]
+  [#local networkVnetResource = networkResources["vnet"]]
+  [#local mgmtSubnetResource = networkResources["subnets"]["mgmt"]["subnet"]]
```
Comment (roleyfoley) [inline components/bastion/setup.ftl]: The network shouldn't be fixed inside of the template and should be determined based on the tier the component is being deployed into

Comment (roleyfoley) [reply components/bastion/setup.ftl]: See the network configuration in https://github.com/codeontap/gen3/blob/581df5b7796f21140b1739eff87794dcf81c2227/providers/aws/components/bastion/setup.ftl#L44 

The get getOccurrenceNetwork is the one which figures this out I'm pretty sure

File: components/bastion/setup.ftl [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+      "PT1M",
+      "Sum",
+      "PT5M",
+      "Average",
+      "LessThan",
+      50000,
+      "Decrease",
+      "ChangeCount",
+      "PT30M",
+      1)]
```
Comment (roleyfoley) [inline components/bastion/setup.ftl]: I'm not sure network bandwidth is a safe way to determine usage of a VM. An ssh session looking at logs can be quite small, especially if there are agents also reporting off to Azure and what not during that time

File: components/bastion/setup.ftl (hunk 2 above)
Comment (RossMurr4y) [reply components/bastion/setup.ftl]: I believe this is the case of an unnecessary variable - networkConfiguration was not used anywhere. Same appears to be the case in the AWS provider.

Nonetheless I've re-arranged these vars - in bc54b00

Comment (roleyfoley) [reply components/bastion/setup.ftl]: Cool, even better

File: components/bastion/setup.ftl (hunk 3 above)
Comment (RossMurr4y) [reply components/bastion/setup.ftl]: As discussed, 3e08cea implements a manual intervention approach for scaling for the bastion MVP. We'll revisit autoscaling for it in time.

**Pull request merged**